# Random Forest Regression Project  

## Tesla Model Y Price Prediction | End-to-End ML Project  

This project focuses on predicting the price of Tesla Model Y cars using **Random Forest Regression**.

It covers the full machine learning workflow, including data preprocessing, exploratory data analysis (EDA), feature selection, and model training.

---

## Dataset  

The dataset used in this project is available on Kaggle: [Dataset Link](https://www.kaggle.com/datasets/aiwithcagri/multiple-linear-regression-tesla-model-y-price)
  

---

## Objective  

The main objective is to build a regression model that can accurately estimate car prices based on key features such as:
- year
- km  
- color 

---

## Workflow  

The project follows these steps:

1. **Data Cleaning**

2. **Exploratory Data Analysis (EDA)**

3. **Feature Engineering** 

4. **Model Building**

5. **Model Evaluation**

---

## Tools & Technologies  

- Python  
- Pandas  
- NumPy  
- Scikit-learn  

---

## Key Learnings  

- Understanding how multiple features affect predictions  
- Importance of proper data preprocessing  
- Handling categorical variables correctly  
- Building and interpreting regression models   

# **Importing The Libaries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data Cleaning Check Template**
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:


In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

# **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv("teslay.csv")
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 255 entries, 0 to 254
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   year    255 non-null    int64  
 1   km      255 non-null    float64
 2   color   255 non-null    object 
 3   price   255 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 8.1+ KB
None

SHAPE OF DATA
Rows: 255, Columns: 4

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 4

DATA TYPES
year       int64
km       float64
color     object
price      int64
dtype: object

NUMERICAL SUMMARY
              year             km         price
count   255.000000     255.000000  2.550000e+02
mean   2024.207843   30384.172549  2.247061e+06
std       0.406562   15563.810709  1.517529e+05
min    2024.000000    6001.000000  1.825000e+06
25%    2024.000000   19880.000000  2.160000e+06
50%    2024.000000   29000.000000  2.239750e+06
75%    2024.000000 

# **Understaning the outliers** 

### km: 6 outliers

In [4]:
# Calculate fences for Price
Q1 = dataset['km'].quantile(0.25)
Q3 = dataset['km'].quantile(0.75)
IQR = Q3 - Q1

lower_fence = Q1 - 1.5 * IQR
upper_fence = Q3 + 1.5 * IQR

# Filter the dataframe to show ONLY the outliers
km_outliers = dataset[(dataset['km'] < lower_fence) | (dataset['km'] > upper_fence)]

# Display the rows
print(km_outliers)

     year        km  color    price
0    2025   69000.0  White  2095000
23   2024   73000.0   Blue  2140000
96   2024   75000.0  White  2010000
145  2024  110000.0  White  2040000
210  2024   96000.0  Black  2049999
229  2024   68200.0  White  2100000


# **Cleaning the data** 

### Remove km outliers (6) & remove the duplication (4)

In [5]:
# Remove rows that exist in km_outliers from the original dataset
dataset_clean = dataset.drop(km_outliers.index).copy()

# Step 2: Remove duplicates FROM the cleaned dataset
dataset_clean = dataset_clean.drop_duplicates().copy()

# Final result
print(dataset_clean)

     year       km  color    price
1    2024  28500.0  Black  2280000
2    2024  50000.0  White  2090000
3    2024  42000.0  White  2180000
4    2024  15000.0  White  2300000
5    2024  36326.0  White  2155000
..    ...      ...    ...      ...
250  2025  24000.0  Black  2390000
251  2024  29000.0  White  2185000
252  2025  19800.0  White  2305000
253  2024  16000.0  White  2250000
254  2024  34000.0  White  2310000

[245 rows x 4 columns]


# **Select Independent Variable (X) and Dependent Variable (y)**

📌 Note: The concepts of Independent and Dependent variables were discussed earlier.

If you would like to review or gain a clearer understanding, please refer to this section: [Go to Independent variable VS Dependent variable Section](https://github.com/Hazem1695/ml-concept-briefs)

In [6]:
# Separate independent variable(features (X)) by selecting all rows and all columns except the last one
X = dataset_clean.iloc[:, :-1].values
# Separate the dependent variable (Target(y)) by selecting the last column
y = dataset_clean.iloc[:, -1].values

# **Encoding categorical data**

## Encoding the Indpendent Variable Using One Hot Encoding

📌 Note: I explained One-Hot Encoding earlier.

If you missed it or want a deeper understanding of how it works, you can find it here: [Go to One-Hot Encoding Section](https://github.com/Hazem1695/ml-concept-briefs)

In [7]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
# One-hot encode the 'color' column (index 2) while keeping other features
# Initialize ColumnTransformer to handle categorical data
# [2] targets the 'color' column (index 2)
ct = ColumnTransformer(transformers = [('encoder', OneHotEncoder(sparse_output=False),[2])], remainder = 'passthrough')
# Execute transformation and convert to NumPy ndarray for model training
X = np.array(ct.fit_transform(X))

# **Splitting the dataset into the Train set and Test set**

In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train , y_test = train_test_split(X, y , test_size = 0.2, random_state = 0)

In [9]:
print (X_test)

[[0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 25000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 22700.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 10124.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 33000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 66000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 41300.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 7650.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 37000.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 44000.0]
 [0.0 0.0 1.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 31000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 19200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 6001.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 19200.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 52531.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2025 41200.0]
 [0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 1.0 2024 29000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2024 25000.0]
 [1.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 2025 41500.0]
 [0.0 1.0 0.0 0

In [10]:
print(f"Rows: {X_test.shape[0]}, Columns: {X_test.shape[1]}")

Rows: 49, Columns: 11


# **Random Forest Regression Model**

## Training the Building Random Forest Regression model on the Training set

In [11]:
from sklearn.ensemble import RandomForestRegressor

# n_estimators = 10: create 10 trees each tree is an estimator

regressor = RandomForestRegressor(n_estimators=100, max_depth=4, min_samples_split=5, min_samples_leaf=3, random_state=0)
regressor.fit(X_train, y_train)

RandomForestRegressor(max_depth=4, min_samples_leaf=3, min_samples_split=5,
                      random_state=0)

1- `max_depth`: The maximum number of levels in the tree.

- Depth = how many splits from root → leaf
- A deeper tree = more complex model

**Effect:**

- Small depth → underfitting (too simple)
- Large depth → overfitting (too complex)

---

2- `min_samples_split`: The minimum number of samples required to split a node.

- The model will NOT split a node unless it has at least this many samples.

**Effect:**

- Small value → more splits → complex tree
- Large value → fewer splits → simpler tree

---

3- `min_samples_leaf`: The minimum number of samples allowed in a leaf (final node)

- After splitting, each leaf must have at least this many samples.

**Effect:**

- Small value → overfitting risk
- Larger value → smoother predictions


# **Preticting the Test set results**

In [12]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[2225521.48 2185000.  ]
 [2230376.27 2250000.  ]
 [2383299.35 2275000.  ]
 [2184862.47 2249990.  ]
 [2110883.5  2089000.  ]
 [2211198.97 2199000.  ]
 [2382103.9  2350000.  ]
 [2463506.07 2425000.  ]
 [2202432.49 2100000.  ]
 [2125331.29 2050000.  ]
 [2269210.07 2320000.  ]
 [2383309.71 2350000.  ]
 [2431966.94 2490000.  ]
 [2346813.89 2365000.  ]
 [2137344.69 2125000.  ]
 [2243917.05 2400000.  ]
 [2215218.19 2265555.  ]
 [2264264.05 2340000.  ]
 [2232495.78 2180000.  ]
 [2271448.06 2350000.  ]
 [2226664.44 2165000.  ]
 [2137956.7  2189000.  ]
 [2233043.8  2185000.  ]
 [2229078.8  2215000.  ]
 [2205618.83 2199000.  ]
 [2219970.47 2240000.  ]
 [2287506.27 2199000.  ]
 [2189418.43 2075000.  ]
 [2431966.94 2500000.  ]
 [2390509.15 2285000.  ]
 [2203047.14 1990000.  ]
 [2217305.48 2130000.  ]
 [2347888.   2300000.  ]
 [2216352.49 2280000.  ]
 [2216549.69 2165000.  ]
 [2203047.14 1950000.  ]
 [2383299.35 2490000.  ]
 [2216130.28 2269900.  ]
 [2120483.8  2195000.  ]
 [2294049.04 2240000.  ]


# **Evaluating the Model Performance**

In [13]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.5440321534037293

# **Important Insight**

1. The initial RandomForestRegressor model produced a low R² score (0.29), despite being a more advanced ensemble method. This indicates that simply using a stronger algorithm does not guarantee better performance.

2. Increasing the number of estimators (n_estimators = 100) and applying regularization constraints (max_depth = 3, min_samples_split = 5, min_samples_leaf = 3) significantly improved performance to R² = 0.53.

3. Further tuning by increasing model capacity slightly (max_depth = 4) resulted in the best performance (R² = 0.54), showing that the model benefited from a better balance between bias and variance.

4. These results confirm that Random Forest performance depends heavily on the quality and configuration of its individual trees. Poorly configured trees lead to weak ensemble performance, while controlled trees produce strong generalization.

5. Across all experiments, it is evident that model performance is primarily driven by:
    - Proper preprocessing (as seen in SVR)
    - Controlled model complexity (as seen in Decision Tree and Random Forest)
    - Careful hyperparameter tuning rather than default settings

---

# **Conclusion**

The experiments demonstrate that machine learning performance is highly dependent on both preprocessing strategy and model configuration, rather than the choice of algorithm alone.

Linear models showed stable but limited performance due to their inability to capture complex relationships. Polynomial Regression improved flexibility but introduced risks of overfitting.

SVR achieved strong performance (R² = 0.53) only after applying selective feature scaling, highlighting the importance of preprocessing decisions.

Tree-based models revealed the importance of controlling model complexity. The Decision Tree model initially failed (R² = -0.35) due to overfitting but improved significantly (R² = 0.49) after applying constraints.

Similarly, the Random Forest model initially underperformed (R² = 0.29) due to a low number of estimators and lack of constraints. After tuning both the number of trees and their complexity, it achieved the best overall performance (R² = 0.54).

These results highlight that effective machine learning requires:

Understanding the relationship between data size and model capacity
Applying selective preprocessing techniques
Systematically tuning hyperparameters

rather than relying on default configurations or model complexity alone.

---

# **Model Comparison**

| Model                 | Configuration                       | R² Score   |
| --------------------- | ----------------------------------- | ---------- |
| Linear Regression     | No Scaling                          | 0.42       |
| Linear Regression     | With Scaling                        | 0.42       |
| Polynomial Regression | With Scaling                        | 0.48       |
| SVR                   | Scale All Features                  | 0.23       |
| SVR                   | Scale Numerical Features (Year, KM) | 0.53       |
| Decision Tree         | Default (Unconstrained)             | -0.35      |
| Decision Tree         | Regularized (Controlled Depth)      | 0.49       |
| Random Forest         | Default (10 Trees)                  | 0.29       |
| Random Forest         | Tuned (100 Trees, Depth = 3)        | 0.53       |
| Random Forest         | Tuned (Depth = 4)                   | **0.54**   |